### PR TITLE
Properly re-print trait impls.

### PR DESCRIPTION
**Pull request by chriseth**
Before this PR, the list of trait implementations was not present in Analyzed, but we need them for re-printing the PIL (and other things).

This PR also contains a simplification in that it only keeps a list of all trait impls, not sorted into which trait they implement. This information is only relevant during trait impl solving, where we re-build that data structure. Keeping the pre-sorted layout would make re-printing in source order much more difficult.

And now it also formats paths to the current namespace properly.

### CODE REVIEW

File: pil-analyzer/src/traits_resolver.rs [hunk 1]
```diff
@@ -52,8 +68,14 @@ impl<'a> TraitsResolver<'a> {
...
             return Ok(());
         };
-        let Some(trait_impls) = self.trait_impls.get(trait_decl_name) else {
+        if !self.traits.contains(trait_decl_name) {
+            // Not a trait function.
             return Ok(());
+        }
+        let Some(trait_impls) = self.trait_impls.get(trait_decl_name) else {
+            return Err(format!(
+                "Could not find an implementation for the trait function {reference}"
```
Comment (gzanitti) [inline pil-analyzer/src/traits_resolver.rs:77]: maybe

```suggestion
                "Could not find an implementation for the trait declaration {reference}"
```

**Review by gzanitti — APPROVED** [review]
LGTM

**Comment by chriseth** [comment]
I'll create a follow-up PR for the error message.

**Comment by chriseth** [comment]
We have a merge conflict, so I'll add the follow-up PR here.

**Review by gzanitti — APPROVED** [review]

**Pull request merged**